infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -292,19 +321,30 @@ public void multiRead(CorfuPayloadMsg<MultipleReadRequest> msg, ChannelHandlerCo
         boolean cacheable = msg.getPayload().isCacheReadResult();
         log.trace("multiRead: {}, cacheable: {}", msg.getPayload().getAddresses(), cacheable);
 
-        ReadResponse rr = new ReadResponse();
+        Lock lock = resetLock.readLock();
+        lock.lock();
+
         try {
-            for (Long address : msg.getPayload().getAddresses()) {
-                ILogData logData = dataCache.get(address, cacheable);
-                if (logData == null) {
-                    rr.put(address, LogData.getEmpty(address));
-                } else {
-                    rr.put(address, (LogData) logData);
+            ReadResponse rr = new ReadResponse();
+            try {
+                if (msg.getEpoch() != serverContext.getServerEpoch()) {
+                    throw new WrongEpochException(serverContext.getServerEpoch());
                 }
+
+                for (Long address : msg.getPayload().getAddresses()) {
+                    ILogData logData = dataCache.get(address, cacheable);
+                    if (logData == null) {
+                        rr.put(address, LogData.getEmpty(address));
+                    } else {
+                        rr.put(address, (LogData) logData);
+                    }
+                }
+                r.sendResponse(ctx, msg, CorfuMsgType.READ_RESPONSE.payloadMsg(rr));
+            } catch (DataCorruptionException e) {
+                r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_CORRUPTION.msg());
```

Comment:
`ERROR_DATA_CORRUPTION.msg()` need to pass a long here. Please pull in #2612

Reply:
MULTIPLE_READ_REQUEST and KNOWN_ADDRESS_REQUEST has a bug, please rebase on #2612

Reply:
#2612 should be merged into corfu-0.3.0.2 otherwise I can't merge into my PR directly

Reply:
So you didn't want to rebase because of the conflicts ? lel

Here's the rebased version #2674

Reply:
Did I say that?
Why you don't want more #2612 into 0.3.0.2 branch?

Reply:
This is part of your patch, and you said you were done. It's reasonable to assume that you would do it, so I wasn't sure why you are waiting for me to cherry-pick it. Anyways, in the interest of time I did it on your behalf.

Reply:
#2612 is another PR, I don't quite see why you want to mix it with this pr, why not just bring it to 0.3.0.2 separately, then I can get it from 0.3.0.2.
Anyways, let's merge #2674 and finish this conversation, don't you agree?